samples/net/ocpp/src/main.c

```diff
@@ -0,0 +1,384 @@
+			  ocpp_io_value_t *io,
+			  void *user_data)
+{
+	static int wh = 6 + NO_OF_CONN;
+	int idx;
+	int i;
+
+	switch (reason) {
+	case OCPP_USR_GET_METER_VALUE:
+		if (OCPP_OMM_ACTIVE_ENERGY_TO_EV == io->meter_val.mes) {
+			snprintf(io->meter_val.val, CISTR50, "%u",
+				 wh + io->meter_val.id_con);
+
+			wh++;
+			LOG_DBG("mtr reading val %s con %d",
+					io->meter_val.val,
+					io->meter_val.id_con);
+			return 0;
+		}
+		break;
+
+	case OCPP_USR_START_CHARGING:
+		if (io->start_charge.id_con < 0) {
+			for (i = 0; i < NO_OF_CONN; i++) {
+				if (!tid[i]) {
+					break;
+				}
+			}
+
+			if (i >= NO_OF_CONN) {
+				return -EBUSY;
+			}
+			idx = i;
+		} else {
+			idx = io->start_charge.id_con - 1;
+		}
+
+		if (!tid[idx]) {
+			LOG_INF("Remote start charging idtag %s connector %d\n",
+					idtag[idx], idx + 1);
+
+			strncpy(idtag[idx], io->start_charge.idtag,
+				sizeof(idtag[0]));
+
+			tid[idx] = k_thread_create(&tinfo[idx], cp_stk[idx],
+						sizeof(cp_stk[idx]), ocpp_cp_entry,
+						(void *)(idx + 1), idtag[idx],
+						obs[idx], 7, 0, K_NO_WAIT);
+
+			return 0;
+		}
+		break;
+
+	case OCPP_USR_STOP_CHARGING:
+		zbus_chan_pub(&ch_event, io, K_MSEC(100));
+		return 0;
+
+	case OCPP_USR_UNLOCK_CONNECTOR:
+		LOG_INF("unlock connector %d\n", io->unlock_con.id_con);
+		return 0;
+	}
+
+	return -ENOTSUP;
+}
+
+static void ocpp_cp_entry(void *p1, void *p2, void *p3)
+{
+	int ret;
+	int idcon = (uint32_t)p1;
+	char *idtag = (char *)p2;
+	struct zbus_observer *obs = (struct zbus_observer *)p3;
+	ocpp_session_handle_t sh = NULL;
+	ocpp_auth_status_t status;
+	const uint32_t timeout_ms = 500;
+
+	ret = ocpp_session_open(&sh);
+	if (ret) {
+		LOG_ERR("ocpp open ses idcon %d> res %d\n", idcon, ret);
+		return;
+	}
+
+	while (1) {
+		k_sleep(K_SECONDS(10));
+		ret = ocpp_authorize(sh,
+				     idtag,
+				     &status,
+				     timeout_ms);
+		if (ret) {
+			LOG_ERR("ocpp auth %d> idcon %d status %d\n",
+				 ret, idcon, status);
+		} else {
+			LOG_INF("ocpp auth %d> idcon %d status %d\n",
+				 ret, idcon, status);
+			break;
+		}
+	}
+
+	if (status != OCPP_AUTH_ACCEPTED) {
+		LOG_ERR("ocpp start idcon %d> not authorized status %d\n",
+				 idcon, status);
+		return;
+	}
+
+	ret = ocpp_start_transaction(sh, sys_rand32_get(), idcon, timeout_ms);
+	if (!ret) {
+		const struct zbus_channel *chan;
+		ocpp_io_value_t io;
+
+		LOG_INF("ocpp start charging connector id %d\n", idcon);
+		memset(&io, 0xff, sizeof(io));
+
+		/* wait for stop charging event from main or remote CS */
+		zbus_chan_add_obs(&ch_event, obs, K_SECONDS(1));
+		do {
+			zbus_sub_wait(obs, &chan, K_FOREVER);
+			zbus_chan_read(chan, &io, K_SECONDS(1));
+
+			if (io.stop_charge.id_con == idcon)
+				break;
+
+		} while (1);
+	}
+
+	ret = ocpp_stop_transaction(sh, sys_rand32_get(), timeout_ms);
+	if (ret) {
+		LOG_ERR("ocpp stop txn idcon %d> %d\n", idcon, ret);
+		return;
+	}
+
+	LOG_INF("ocpp stop charging connector id %d\n", idcon);
+	k_sleep(K_SECONDS(1));
```

Comment:
These sleeps here and below looks a bit problematic. They probably work ok now but could fail in the future if something changes in the system. If you need to wait something to finalize, better would be to have a semaphore that we wait here. We have used that kind of pattern in other networking samples.